### PR TITLE
[CI] Run code coverage in nightly

### PR DESCRIPTION
**Pull request by sausagee**
## Motivation
We need to track the code coverage to make sure the basic testing is in place to exercise all the code.

We are adding grcov to the CI pipeline to collect and aggregate code coverage. The coverage result is uploaded to codecov.io. The dashboard can help visualize the breakdown of test coverage.

## Test Plan
Canaried the code-coverage flow with 02b7b6a in https://circleci.com/gh/libra/libra/16998?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link
image

The coverage report lcov.info is uploaded to codecov.io
https://codecov.io/gh/libra/libra/tree/02b7b6a13c703c96c13724efe409a43fb78ede6f
image

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -109,6 +109,49 @@ jobs:
...
+          command: rustup install nightly
+      - run:
+          name: Install grcov
+          command: cargo install --force grcov
+      - run:
+          name: Setup rustc flags
+          command: |
+            echo "export CARGO_INCREMENTAL=0" >> $BASH_ENV
+            echo "export RUSTFLAGS='-Zprofile -Ccodegen-units=1 -Cinline-threshold=0 -Clink-dead-code -Coverflow-checks=off -Zno-landing-pads'" >> $BASH_ENV
+      - run:
```
Comment (bob-wilson) [inline .circleci/config.yml]: Is it necessary to first do a "cargo build"? The subsequent "cargo test" will rebuild at least most of the code anyway (AFAIK) and the "cargo build" will compile some code that cannot be exercised in the tests. Unless there's a good reason, I think you should skip this.

File: .circleci/config.yml [hunk 2]
```diff
@@ -109,6 +109,49 @@ jobs:
...
+          name: Cargo build
+          command: cargo +nightly build
+      - run:
+          name: Cargo test
```
Comment (bob-wilson) [inline .circleci/config.yml]: I'm not sure what the "+nightly" option does here. What tests get included in this?

File: .circleci/config.yml [hunk 3]
```diff
@@ -109,6 +109,49 @@ jobs:
...
+          command: cargo +nightly test
+      - run:
+          name: Generate coverage report
+          command: |
+            zip -0 ccov.zip `find . \( -name "libra*.gc*" \) -print`
```
Comment (bob-wilson) [inline .circleci/config.yml]: Why only "libra*.gc*"? What files are you excluding with that pattern?

**Comment by sausagee** [comment]
@bob-wilson I need to run a few more canaries to validate the work flow.  I am going to remove you from the reviewer list to avoid spamming you with noise.  Will add you back once the PR is out of draft mode.

File: .circleci/config.yml (hunk 1 above)
Comment (sausagee) [reply .circleci/config.yml]: Good point. The build step generates additional artifacts we need to exclude or they end up in the report.  Will patch this up.

File: .circleci/config.yml (hunk 2 above)
Comment (sausagee) [reply .circleci/config.yml]: The current default toolchain is beta, but it doesn't support -Zprofile. So we are telling cargo to use the nightly toolchain.

File: .circleci/config.yml (hunk 3 above)
Comment (sausagee) [reply .circleci/config.yml]: Good question!  I don't know what is missing yet.  I am trying to plumb the work flow first and put the results on a dashboard making it easy to digest.

**Comment by sausagee** [comment]
Rebased due to recent change in build that changed how cargo test is run.  

Currently canarying the flow on CircleCI.

**Comment by sausagee** [comment]
Looks like the plumbing works.  Just need to figure out the 0 coverage next.  When tested locally with grcov's recipe, I saw 84% coverage. 

I updated the work flow with `scripts/coverage_report.sh` but it generated a report with 0 coverage.  @bob-wilson do you see anything obvious that would cause it?  At the meantime, I will try to get the raw lcov.info from circle (if not, I will repro locally)

**Comment by bob-wilson** [comment]
I don't see anything obviously wrong. Maybe this will be fixed by https://github.com/libra/libra/pull/1523?

**Comment by sausagee** [comment]
Somehow the log in Circle's build didn't capture the stderr.

Repro-ed the issue on my dev server with 0 coverage. Don't know yet why the checksum would be different.
```
test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests client

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

Generating lcov report at codecov/lcov.info...
22:33:06 [ERROR] Error in computing counters: File checksums do not match: 1386628098 != 1017764555 in debug/deps/miniz_oxide-a73c908dcf1a65db
22:33:06 [ERROR] Error in computing counters: File checksums do not match: 613053570 != 1416630387 in debug/deps/prost_derive-1d9dbcd2beb2e861
22:33:06 [ERROR] Error in computing counters: File checksums do not match: 3771400052 != 783098701 in debug/build/failure_derive-5a59685ba3278008/build_script_build-5a59685ba3278008
22:33:06 [ERROR] Error in computing counters: File checksums do not match: 530971125 != 3919636199 in debug/build/syn-7dc90910d41d0cd9/build_script_build-7dc90910d41d0cd9
22:33:06 [ERROR] Error in computing counters: File checksums do not match: 1686303153 != 1490043857 in debug/build/bindgen-981e0319cceca5d3/build_script_build-981e0319cceca5d3
22:33:06 [ERROR] Error in computing counters: File checksums do not match: 2879278914 != 2286487459 in debug/deps/same_file-716c7a8e8485c074
22:33:06 [ERROR] Error in computing counters: File checksums do not match: 4171309246 != 52740952 in debug/deps/proc_macro2-5427030bd3b277e0
22:33:06 [ERROR] Error in computing counters: File checksums do not match: 3276392921 != 603037411 in debug/deps/nibble_vec-cba0390f6cba99a0
...
...
Processing file language/transaction-builder/src/lib.rs
genhtml: WARNING: cannot read language/transaction-builder/src/lib.rs!
Writing directory view page.
Overall coverage rate:
  lines......: 0.0% (0 of 5463 lines)
  functions..: 0.0% (0 of 718 functions)
Done. Please view report at codecov/index.html
```

**Pull request closed without merging**